### PR TITLE
goal-conditioned sampler API update

### LINKED ISSUE #422: allow samplers to be goal-conditioned

### PR DESCRIPTION
**Pull request by tomsilver**
closes #422

### CODE REVIEW

File: src/datasets/demo_replay.py [hunk 1]
```diff
@@ -55,6 +55,7 @@ def create_demo_replay_data(env: BaseEnv,
         # Sample a trajectory
         traj_idx = rng.choice(len(demo_dataset.trajectories), p=weights)
         traj = demo_dataset.trajectories[traj_idx]
+        goal = train_tasks[traj.train_task_idx].goal
```
Comment (ronuchit) [inline src/datasets/demo_replay.py:58]: lol i would have definitely messed up here and used `train_tasks[traj_idx]` which would have worked for now but possibly caused horrible bugs later in life

File: src/ground_truth_nsrts.py [hunk 2]
```diff
@@ -400,7 +411,7 @@ def grasp_sampler(state: State, rng: np.random.Generator,
         }
         add_effects = {LiftedAtom(HandEmpty, [])}
         delete_effects = {LiftedAtom(Holding, [can])}
-        place_sampler = lambda s, r, o: np.array([
+        place_sampler = lambda s, g, r, o: np.array([
```
Comment (ronuchit) [inline src/ground_truth_nsrts.py:414]: wasn't the goal of this PR to make a goal-conditioned version of the place sampler? is that coming in a separate PR?

**Review by ronuchit — APPROVED** [review]
whew, this is an awesome and large change! great stuff.

File: src/ground_truth_nsrts.py (hunk 2 above)
Comment (tomsilver) [reply src/ground_truth_nsrts.py:414]: yeah that will happen separately

**Pull request merged**